leaderelection/leaderelection.go

```diff
@@ -35,82 +35,82 @@ import (
+	Stop()
 }
 
 type Manager struct {
-	log    *logp.Logger
-	leader *le.LeaderElector
-	block  chan bool
+	log        *logp.Logger
+	leader     *le.LeaderElector
+	wg         *sync.WaitGroup
+	cancelFunc context.CancelFunc
+	kubeClient k8s.Interface
 }
 
-// NewLeaderElector acts as a singleton
-func NewLeaderElector(ctx context.Context, log *logp.Logger, cfg config.Config) (ElectionManager, error) {
-	// the launcher is creating a new beater instance every new configuration.
-	// therefore, we must reset callOnce every new beater to stall the fetching cycle until we have the leader information.
-	*callOnce = sync.Once{}
+func NewLeaderElector(log *logp.Logger, cfg config.Config) (ElectionManager, error) {
 	kubeClient, err := providers.KubernetesProvider{}.GetClient(cfg.KubeConfig, kubernetes.KubeClientOptions{})
 	if err != nil {
 		log.Errorf("NewLeaderElector error in GetClient: %v", err)
 		return nil, err
 	}
 
-	initOnce.Do(func() {
-		var leConfig le.LeaderElectionConfig
-		var leader *le.LeaderElector
-
-		block := make(chan bool)
-		leConfig, err = buildConfig(ctx, log, kubeClient, block, callOnce)
-		leader, err = le.NewLeaderElector(leConfig)
+	wg := &sync.WaitGroup{}
 
-		manager = &Manager{
-			log:    log,
-			leader: leader,
-			block:  block,
-		}
-	})
-
-	return manager, err
-}
-
-func GetLeaderElectorManager() ElectionManager {
-	return manager
+	return &Manager{
+		log:        log,
+		kubeClient: kubeClient,
+		leader:     nil,
+		cancelFunc: nil,
+		wg:         wg,
+	}, nil
 }
 
 func (m *Manager) IsLeader() bool {
 	return m.leader.IsLeader()
 }
 
-// Run leader election is blocking until a leader is being elected or timeout has reached.
+// Run leader election is blocking until a FirstLeaderDeadline timeout has reached.
 func (m *Manager) Run(ctx context.Context) error {
-	timeoutCtx, cancel := context.WithTimeout(ctx, FirstLeaderDeadline)
-	defer cancel()
+	newCtx, cancel := context.WithCancel(ctx)
+	m.cancelFunc = cancel
 
-	go m.leader.Run(ctx)
-	m.log.Infof("started leader election")
+	leConfig, err := m.buildConfig(newCtx)
+	if err != nil {
+		m.log.Errorf("Fail building leader election config: %v", err)
+		return err
+	}
 
-	select {
-	case <-m.block:
-		m.log.Infof("new leader has been elected")
-	case <-timeoutCtx.Done():
-		m.log.Warnf("timeout - no leader has been elected for %s", FirstLeaderDeadline.String())
+	m.leader, err = le.NewLeaderElector(leConfig)
+	if err != nil {
+		m.log.Errorf("Fail to create a new leader elector: %v", err)
+		return err
 	}
 
+	go m.leader.Run(newCtx)
+	m.wg.Add(1)
+	m.log.Infof("started leader election")
+
+	time.Sleep(FirstLeaderDeadline)
+	m.log.Infof("stop waiting after %s for a leader to be elected", FirstLeaderDeadline)
+
 	return nil
 }
 
-func buildConfig(ctx context.Context, log *logp.Logger, client k8s.Interface, block chan bool, callOnce *sync.Once) (le.LeaderElectionConfig, error) {
-	podId, err := currentPodID(log)
+func (m *Manager) Stop() {
+	if m.cancelFunc != nil {
```

Comment:
Add error log in case `cancelFunc` not set

Reply:
Nitpicking, I would invert the `if` statement (so the more interesting part is not indented in)